### PR TITLE
Refactor kube-scheduler deployment

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area control-plane dev-productivity quality robustness testing
/kind enhancement
/priority normal
/topology seed
/platform all
/exp intermediate

**What this PR does / why we need it**:
This PR refactors the deployment for the kube-scheduler and adopts the `DeployWaiter` interface for components. Consequently, it eliminates the Helm chart and adds full unit tests, instead. This improves several aspects and the overall code quality.

Part of #2754

**Special notes for your reviewer**:
I haven't touched the `kube-scheduler` related system resources (https://github.com/gardener/gardener/tree/master/charts/shoot-core/components/charts/kube-scheduler) but could follow up in a subsequent PR with it.

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/controlplane/kube_scheduler.go
Comment (timuthy) [inline pkg/operation/botanist/controlplane/kube_scheduler.go]: Do you also want to add v1beta1constants.LabelPodMaintenanceRestart: "true"? I already wanted to do it in another PR but then I saw this one 😄

Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kube_scheduler.go]: Sure, I can do it in a follow-up PR once this one is merged (to not mix a refactoring with a new feature) 👍

Comment (timuthy) [reply pkg/operation/botanist/controlplane/kube_scheduler.go]: Okay, thank you :)

**Comment by ialidzhikov** [comment]
To be honest, I am with mixed feelings wrt helm vs go struct approach. Some cons against the go struct approach that I see at first sight:
- I like that currently with helm each manifest is in its own separate file. I usually search for given deployment or other manifest and I can find it fast when I know the naming convention of the files. With the go struct approach all objs are `CreateOrUpdate`'d in single `Deploy` func which makese harder to maintain I think.
- the go struct approach seems too verbose. I see that multuple computations are extracted into smaller funcs (for example `computeCommandLineFlags`) which leads to going up and down in the source - for example need to check utility func at L371, then go back to L130 where this partial spec is stored in variable and then go L199 to check how this variable is used.
- `computeVersionConstraints` func alternative in helm comes for free
- with helm we don't need to vendor `k8s.io/kube-scheduler`
- go struct way does not scale and it is not maintainable for huge charts

Probably you can share what are the pros? Probably easier testing? But it should be possible to achieve that also with helm + chartRenderer and chartApplier, right?

/cc-ing also  @gardener/gardener-maintainers for feedback

**Comment by rfranzke** [comment]
> With the go struct approach all objs are CreateOrUpdate'd in single Deploy func which makese harder to maintain I think.

Just to understand it correctly, your point here is that the deployment is not in its own file, right? If it's only that, I'm open to move the kube-scheduler related things into its own package and have a dedicated file per resource, just like in the Helm chart.

> the go struct approach seems too verbose. 

Yes, there is some boilerplate, agreed.

>  I see that multuple computations are extracted into smaller funcs (for example computeCommandLineFlags) which leads to going up and down in the source - for example need to check utility func at L371, then go back to L130 where this partial spec is stored in variable and then go L199 to check how this variable is used.

I don't think this is a valid argument. You'll find lots of `{{ template }}` in the Helm charts that have more complex logic as well, which are later ` {{ include }}`d at the right place in the actual manifest. That's exactly the same, we just haven't followed this approach extensively everywhere (and, hence, have only a few `{{ if-then-else }}` expressions which make the impression that it's "easier to read").

> computeVersionConstraints func alternative in helm comes for free

Yes, but to the price that errors are not handled properly and might cause side-effects if certain conditions don't hold true. That's exactly what we are trying to improve now (although, this particular error handling of the version comparisons is not the main motivation, of course).

> with helm we don't need to vendor k8s.io/kube-scheduler

Well, so what? We only vendor it for the APIs of the configuration. I don't see any negative consequences doing so. What is your point here?

> go struct way does not scale and it is not maintainable for huge charts

For some things it might make sense to keep them (CRDs, for example), sure. I don't suggest to get rid of the Helm chart/chart applier/etc. code in its entirety.

As discussed in today's meeting, the main motivation for this PR was to improve the code coverage wrt unit tests. Using the already established testing frameworks with Ginkgo/Gomega and Gomock was the most straight-forward way without the necessity of introducing new concepts/complexities.
The current structure with the charts, their (default) values, and the corresponding Go implementations is too fragile and error-prone. Too many parts are distributed, there is no consistent data flow, and both cause a high risk of breaking something as we don't have the tests.

I'm not completely against the Helm charts and the chart applier. An alternative approach would be to explicitly provide the important values in the Go code for rendering the chart (like deployment name, labels, container name, ports, etc.) and don't "hard-code" them in the chart itself. Then we can have unit tests to check if the correct Helm values are computed by Go, but we'll also need other unit tests for the Helm chart itself to check if the values are used in the correct way (e.g., that the container name lands in `.spec.template.spec.containers[].name`), and so on. What will remain with this approach is the distribution of the code (you won't have one place in the tree where everything related to kube-scheduler is, you'll still have the chart and the Go code). How would the chart then look like in the end? It won't be that "readable" as it's toady, that's for sure. In that shape, is it still valuable enough to justify the complexity?
Also, we would need tests for the various `if-then-else` constructs in the Helm charts. And how would such a test look like? Rendering it and then comparing the result against a pre-defined YAML file with the expected output? Sure, can be done, but to be honest, isn't this more complicated and doesn't solve our problems today? I think, already today have charts where we have differences for the various Kubernetes versions, so we have `if-then-else` constructs in the chart, but then we also have them in the Go code to provide the correct values. This is totally unnecessary and hard to read/understand.

Do you have other ideas/proposals? I'm not insisting on this approach, there might be better ways, but I like that it solves the problems of testability, code distribution, and easier data flow (this part is not yet consistent, of course, but will improve over time). As @timuthy suggested, we can put the whole kube-scheduler related code into its own package, this will probably help a little bit more. However, I'm open to better approaches and very curious of what @timebertt will present end of the week.

**Comment by mvladev** [comment]
I think that this PR is setting things in the right direction. I would like to have a separate controller for each component (`kube-scheduler`, `kube-apiserver` and so) and this is an intermediate solution until then.

It's a little more verbose, but it can be made many times more testable than the current solution. The boilerplate code could be refactored later on as common code is extracted into packages / structs.

File: pkg/operation/botanist/controlplane.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane.go]: Can you change the return type of `GetReplicas` to `int32`?
```suggestion
		b.Shoot.GetReplicas(1),
```

File: pkg/operation/botanist/controlplane.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane.go]: Can we have a separate constant for the name of the kubeconfig secret to avoid confusion/mistakes?
Maybe it makes sense to have those constants in the kubescheduler package or should we have them in the common package?

Also, can you please pass the respective constants to the `Name` field, as it will be used as the `SecretRef` in the Volume?
```suggestion
		Kubeconfig: kubescheduler.Secret{Name: v1beta1constants.SecretNameKubeScheduler Checksum: b.CheckSums[v1beta1constants.SecretNameKubeScheduler]},
		Server:     kubescheduler.Secret{Name: common.KubeSchedulerServerName, Checksum: b.CheckSums[common.KubeSchedulerServerName]},
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go:55]: Just a question: Am I assuming correctly that this constant should be used by the prometheus component later on to compute the scrape config for the kube scheduler? Is this the concept for how we want to improve the data flow between components?

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Nit, this doesn't only compute flags, but rather the whole command 😉 
```suggestion
func (k *kubeScheduler) computeCommand(port int32) []string {
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Can we remove those defaults?

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Just a suggestion :)
Can be applied to all the other paths (in the command line flags) too, if you want.
```suggestion
			Kubeconfig: filepath.Join(volumeMountPathKubeconfig, secrets.DataKeyKubeconfig),
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: I'm thinking about how this computation and the `kubeSchedulerVersionConstraints` construct can be simplified to reduce the amount of explicit error handling. Also this will parse the Shoot version for every constraint and each constraint for every shoot being reconciled...

WDYT about:
- "pre-computing" the semver constraints in an `init` func (panicking on error, as this is a programmer level error, maybe with `utilruntime.Must`)
- and parsing the shoots version to a `*semver.Version` when constructing the `Shoot` operation object?

Constraints can then be checked by just calling the `Check` func (which doesn't return an error).
This would slim down this component and the `New` func also wouldn't need to return an error. (Also we would have implicit test coverage for the constraints 😉 )

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: Can you put that into a multi-line string to make it easier to read?

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]:
```suggestion
	if k8sVersionGreaterEqual113, _ := versionutils.CompareVersions(version, ">=", "1.13"); k8sVersionGreaterEqual113 {
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: Any specific reason, why you use a literal for `TerminationMessagePath` and a constant for `TerminationMessagePolicy`?

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]:
```suggestion
											Name:      configMapName,
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]:
```suggestion
									Name: configMapName,
									VolumeSource: corev1.VolumeSource{
										ConfigMap: &corev1.ConfigMapVolumeSource{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
```

File: pkg/operation/botanist/controlplane.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane.go]: I think the cached client should be enough for this case. Have you tested this and encountered any issues with it?
```suggestion
		b.K8sSeedClient.Client(),
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: Can you use the `DeepDerivativeEqual` to get the enhanced YAML diff?
```suggestion
						c.EXPECT().Update(ctx, gomock.AssignableToTypeOf(&corev1.ConfigMap{})).
							Do(func(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) {
								Expect(obj).To(DeepDerivativeEqual(configMapFor(version)))
							}),
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go:336]: On 1.18 the scheduler crashes with the following error:
```
: v1alpha2.KubeSchedulerConfiguration.ClientConnection: v1alpha1.ClientConnectionConfiguration.BindTimeoutSeconds: TypeMeta: APIVersion: ReadObject: found unknown field: algorithmSource, error found in #10 byte of ...|thmSource":{},"apiVe|..., bigger context ...|{"algorithmSource":{},"apiVersion":"kubescheduler.config.k8s.io/v1al|...
```

The reason seems to be, that `algorithmSource` was removed in [`kubescheduler.config.k8s.io/v1alpha2.KubeSchedulerConfiguration`](https://github.com/kubernetes/kube-scheduler/blob/da23134c2b6a206bfa461e2779b32c7c93d19708/config/v1alpha2/types.go#L40).

Is there a possibility to not marshal the fields, that we don't explicitly set? (I see a lot of fields in the config API don't have `omitempty` tags). Maybe we should go back to a go template for the config 🤔

**Review by timebertt — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks good in general, I somehow like the go struct approach, now that I have spent some time with it in the review :)

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (stoyanr) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: I would inject the codec, instead of creating it here. This would allow for it to be easily mocked away in tests, to be replaced by a different implementation, etc.

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go:55]: Yes, I am working on a subsequent PR that improves those points, right.

**Review by stoyanr — COMMENTED** [review]
Just a small comment regarding the codec creation and usage that @timebertt brought to my attention.

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Wouldn't this break if you develop/build Gardener on e.g. Windows?

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: I don't use the constants in the tests on purpose to detect potential issues when they get changed.

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: see https://github.com/gardener/gardener/pull/2731#discussion_r476267068

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go:336]: Yes, I think it makes sense to go back to a template for the configuration. It is already cumbersome to overwrite those API versions, and it will become more cumbersome when the structure differs between the version (e.g., for the `algorithmSource` field).

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Thanks, I think we will go back to a template for the configuration (at least for the kube-scheduler), see https://github.com/gardener/gardener/pull/2731#discussion_r475690713.

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go:55]: Ok perfect, makes sense to me.

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go
Comment (timebertt) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler.go]: Oh well, I missed that. Just forget my suggestion, then 😄

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: That makes sense in general. But, `configMapName` is not a constant from the production code, it's only a variable in your test. I think, there's no problem with reusing that :)

Comment (rfranzke) [reply pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: Ah, got it, yep, thanks

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]: Nit
```suggestion
					Name:       deploymentName,
```

File: pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go
Comment (timebertt) [inline pkg/operation/botanist/controlplane/kubescheduler/kube_scheduler_test.go]:
```suggestion
												Name: configMapName,
```

**Review by timebertt — COMMENTED** [review]
Thanks for incorporating my feedback!
Looks great now, tested it again and works with 1.17 and 1.18, now.

Only two small suggestions to make the test more consistent and easier to adapt in the future, but no strong opinion there.

**Review by timebertt — APPROVED** [review]
Thanks! Good to go from my POV 🚀 
/lgtm
/approve

**Pull request merged**
